### DISCUSSION TITLE
How to use Equilizer3Bands?

### DISCUSSION THREAD

**Discussion by timeylies** [pschatzmann/arduino-audio-tools#227] [Q&A] [answered]
Does anybody know how to use Equilizer3Bands, if so please tell me how to use it?

**Comment by pschatzmann** [comment]
You use the Equilizer3Bands as ouput class indicating the final output class in the constructor:

e.g. 
```
...
I2SStream i2s;
Equilizer3Bands eq(i2s);
StreamCopy copier(eq, in); // copy in to equilizer
...
```
You configure the Equilizer like any other object:
```
...
auto cfg = eq.defaultConfig();
cfg.gain_low = 0.5
cfg.gain_medium = 0.5
cfg.gain_high = 1.0
eq.begin(cfg);
...
```
You might also need to set the channels, bits_per_sample or sample_rate e.g. with eq.setAudioInfo(is2_cfg); or by assigning the values.
The gain values can be updated dynamically in the loop.

Please note that I have never tested this class, so any feed-back is welcome...

**Reply by timeylies** [reply]
Thanks!

**Reply by timeylies** [reply]
I tried that but the error says that: **_cannot declare variable 'eq' to be of abstract type 'audio_tools::Equilizer3Bands'_**
I'm using the BluetoothA2DPSink and AudioKitStream for input and output, and I want the equalizer to be between the input and output like a filter

**Comment by pschatzmann** [comment, marked as answer]
I can reproduce your issue: I have committed a correction and it is compiling now w/o errors...

**Reply by timeylies** [reply]
It does compile but when my esp32 loads it, it keeps resetting
I do believe that it is resetting because I did:  
`// Write A2DP data to eq in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
  eq.write(data, length);
}`

**Reply by timeylies** [reply]
What I wrote above wasn't the reason, it probably was that I used the AudioKitStream as a stream for the eq, which was like **_Equilizer3Bands eq(kit)_**

**Reply by pschatzmann** [reply]
The next correction has been committed. Here is an [example sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/basic-a2dp-eq-audiokit/basic-a2dp-eq-audiokit.ino)

**Comment by cQszasza** [comment]
Hi! Is it possible to change the low/high frequency settings during runtime? I can only change the weight and that works fine but when I try to change the frequencies nothing happens. I can change them by re uploading the code with different values, but not when running. Any ideas?

**Reply by pschatzmann** [reply]
Mybe you just forgot to call eq.begin(cfg); with the new settings ?

**Reply by cQszasza** [reply]
Thanks, I'll try it. So, for the weights I don't have to do that but for the frequencies I do? Is there a reason behind this?

**Reply by cQszasza** [reply]
Yep, calling begin works, thank you!

**Reply by pschatzmann** [reply]
Yes, I suggest that you have a look at the implementation. After all it is just 233 lines of code: Changing the frequencies requires a recalculation of the EQSTATE which is quite expensive.

**Comment by vinhx** [comment]
Hi Sir,
I use the example "bt_music_receiver_simple" from your library, run it with my ESP32 Dev Module that use ESP32-S chip, it run smoothly, my idea is to add the Equilizer3Bands into this and control the values from some button, I try to read your sample "basic-a2dp-eq-audiokit.ino"
but I have to admit that I just new in Arduino with with limitation on programming, I couldn't find out how to understand the AudioKit.h library in backwards to understand how to get the settings for my board instead,
could you please help to give some tip, thank you a lot!

**Comment by pschatzmann** [comment]
The AudioKitStream is the input/output class used e.g. by the AI Thinker AudioKit board or ES8388 based solutions. So this not for your use case!

Just use the I2SStream instead of the AudioKitStream and you are good to go.

Don't forget the default pin assignments used by the A2DP library and  I2SStream are different: https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC 

ps. next time look at the README that you can find in the correspondig examples directory: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit

**Reply by vinhx** [reply]
thank you Sir!
I will try with your advice
